[BUGFIX] enable lazy-relationship payloads to work with polymorphic relationships

runspired opened the pull request:
*Once reviewed/approved, this should be backported to 2.14, 2.15, and 2.16 beta.*

Resolves #5037 #5055 #5208 #5209 #5211 #5267 #5268 

**Problem**

Ember Data 2.14 introduced lazy-relationship parsing; however, this parsing used left-side/right-side keying that was incompatible with polymorphic relationships.  This resulted in payloads for newly encountered polymorphic types causing all previous payloads to be dereferenced.

In addition, a number of bugs were discovered where we were incorrectly setting up `RelationshipPayloads` as if there was an inverse when there was none, or where the inverse resolved to the key `:`.  This PR fixes that from the perspective of RelationshipPayloads but part of the problem was a bug in our calculation of `inverseFor` which is resolved with https://github.com/emberjs/data/pull/5261

**Solution**

This PR resolves these issues with three primary changes:

1) `relationship-info` is now [calculated and stored separately](https://github.com/emberjs/data/pull/5230/files#diff-1c53e096d92ba3a647478f3f51185a01R182) from the `RelationshipPayloads` objects.  This not only should be a performance win for us by enabling us to avoid calculating out the info in several scenarios where we would have needed to repeatedly redo the work before, but helps to resolve lookup for scenarios where there is no inverse or the inverse is polymorphic.  Additionally it cleans up many areas of the lazy-relationship code substantially to have this separation of concerns.

2) `RelationshipPayloads` are now always [cached on the manager by the resolved `base-model-name` instead of `model-name`](https://github.com/emberjs/data/pull/5230/files#diff-1c53e096d92ba3a647478f3f51185a01R173).  This enables us to confidently always utilize the correct `RelationshipPayloads` even when encountering a polymorphic type.

3) The payloads held for a relationship in `RelationshipPayloads` are now stored in [a two-level cache by `model-name` and `id`](https://github.com/emberjs/data/pull/5230/files?w=1#diff-ce1590fc9040b7fec82c5c59304875adR82) instead of a single-level cache of just `id`. This enables situations in which polymorphic siblings do not share a common unique-id constraint to work properly.

**Tests**

- one-side polymorphic belongsTo / hasMany
- both-sides polymorphic belongsTo / hasMany
- multi-table polymorphism (e.g. when IDs are not unique across all polymorphic types)
- reflexive polymorphic relationships
- self-referential non-reflexive polymorphic relationships

hjdivad commented on addon/-private/system/relationships/relationship-payloads.js:
```diff
@@ -44,10 +44,32 @@ export default class RelationshipPayloads {
+      inverseRelationshipMeta.options !== undefined &&
+      inverseRelationshipMeta.options.polymorphic === true;
+
+    assert(`Both side of a relationship cannot be polymorphic`, !this._lhsIsPolymorphic || !this._rhsIsPolymorphic);
```
as we discussed out of band, it's fine for both sides to be polymorphic

hjdivad commented on addon/-private/system/relationships/relationship-payloads-manager.js:
```diff
@@ -205,6 +209,22 @@ export default class RelationshipPayloadsManager {
     let lhsKey = `${modelName}:${relationshipName}`;
     let rhsKey = `${inverseModelName}:${inverseRelationshipName}`;
```
as discussed out of band, in both cases we want the key to be the model name defined by the inverse relationship, when an inverse relationship exists

hjdivad replied:
so we should always create `RelationshipPayloads` with the base name and then add the actual model name

hjdivad commented on addon/-private/system/relationships/relationship-payloads.js:
```diff
@@ -44,10 +44,32 @@ export default class RelationshipPayloads {
   constructor(store, modelName, relationshipName, relationshipMeta, inverseModelName, inverseRelationshipName, inverseRelationshipMeta) {
```
maybe s/`modelName`/`baseModelName` or something here as we want this to always be the base model name

hjdivad commented on addon/-private/system/relationships/relationship-payloads.js:
```diff
@@ -126,14 +159,28 @@ export default class RelationshipPayloads {
     }
   }
 
+  _isSide(potentialMatches, modelName, relationshipName) {
```
👍

hjdivad commented on tests/unit/system/relationships/relationship-payload-manager-test.js:
```diff
@@ -674,3 +678,72 @@ test('get can retrieve payloads with self-links in reflexive relationships', fun
   assert.deepEqual(entry, friendsPayload, 'self-link in reflexive relationship');
 });
+
+test('handles relationships where one side is polymorphic', function(assert) {
```
as discussed out of band, need more tests

- push subtype then push basetype
- push basetype then push subtype
- both sides poly

danielspaniel replied:
Any way I can keep this moving?  Is it possible for me to merge tests into this feature branch? Never tried that before

hjdivad replied:
@danielspaniel you can open pull requests against runspired/data to branch `fix-polymporphic-payloads` and when they're merged it'll update this pr.  I've done it before and it works surprisingly well.

@danielspaniel if you have time to contribute tests for this that would be super helpful.  Thanks ^_^

danielspaniel replied:
I have all tests ready but @runspired  added this to RelationshipPayloads constructor: 
```
assert(`Both side of a relationship cannot be polymorphic`, !this._lhsIsPolymorphic || !this._rhsIsPolymorphic);
```

That is showstopper, since previously in ember data you could do that .. so I think that assertion is perhaps guarding agains something I don't know what.  

But I removed that assertion and the test failed for both sides polymorphic, so maybe that had something to do with it .. not sure

danielspaniel replied:
This feature ( lazy relationship loading ) is super complex and was implemented with very limited tests in the first place.  
 I am thinking it might be better to dump the feature,  rather than keep patching it.

  Was this your feature @hjdivad ?  Is there a great value to it.  If there is I am not seeing it. 
Would you be up for dumping it/ removing it?

danielspaniel replied:
So, if a feature is not done properly and introduces very serious bugs,  are you telling me you are reluctant to back out the feature because you would lose the nice performance gains? 
 That makes no sense. 
First off, whoever wrote the feature should take responsibility and either fix it or remove it. 
I am willing to help out as much as I can, but the way it was done really looks like spaghetti to me, so I can't do much to save it. 
Secondly, if I wrote this feature, I would be embarrassed to keep it, if it broke ember-data. Does not really matter what kind of fun it offers if it is broken.  

How long should this broken feature sit in ember-data and just kill off polymorphic relationship users? 
Is it right to just wait around for some other work someone else is doing?  Who knows how long that other thing Igor is working on will be done.  Seems irrelevant to me, especially since he is not involved in this discussion much. It has only been you, me , Chris and Stephan ( saying he liked the original bug posting  )

runspired replied:
@danielspaniel Sorry for being a bit quiet. Last week when @hjdivad and I discussed this PR OOB we discussed that assert.  Polymorphic to Polymorphic in ED has always been very buggy, and I wasn't sure it really worked at all and I thought it'd be too difficult to achieve here; however, in that conversation David and I also arrived at a solution for supporting it. So long as we always resolve to the baseModelName on both sides, the approach this PR outlines will work for both sides being polymorphic as well.  I will update the PR with that approach.

runspired replied:
@danielspaniel can you PR the tests you've added to my branch?

danielspaniel replied:
sure thing .. gimme a few moments

danielspaniel replied:
I submitted PR to your branch of ED just now

hjdivad commented:
@runspired possibly also #5208

runspired commented:
Updated to handle polymorphic-to-polymorphic relationships and with the help of @danielspaniel we now have more tests. I would like to clean up the testing a bit more but this is ready for another review pass by @hjdivad.

The only piece that concerns me is that we now MUST lookup the inverse greedily in a situation we did not need to before here: https://github.com/emberjs/data/pull/5230/files#diff-1c53e096d92ba3a647478f3f51185a01R178

I'm also sure now that we've got an implementation that covers the appropriate cases that this is due for some further refactoring and simplification.

BryanCrotaz commented:
Haven't looked at source in detail but have you tested with multiple levels of model extension?

E.g. A extends B extends C, with relationships introduced at each level?

runspired commented:
@BryanCrotaz those cases introduce no additional complexity or scenarios. A relationship must explicitly declare that it is polymorphic, and it must use the base type's type when doing so.

BryanCrotaz commented:
@runspired :) will try it out

danielspaniel commented:
This fix is still not complete enough for polymorphic relationships.  I found that  polymorphic self referential relationships  don't work  .. 2 options 

 - I can add the failing tests to this branch, @runspired can send more brainwaves at the problem  
 or 
 - we can just give up now, rip out this lazy relationships manager and save everyone who loves polymorphism   from dying an early death from alcoholism

runspired commented:
Another failing test please

danielspaniel commented:
I guess you beat me to it .. I was working on it but you added a whole ton of self-reflexive tests that were the ones that were breaking for me

danielspaniel commented:
actually check out my test .. it still fails  ( and I just rebased )

runspired commented:
I'm afk but reading those asserts I believe they assert the incorrect thing.

danielspaniel commented:
not sure why you think that .. but let me know when you get a chance.

runspired commented:
@danielspaniel ported and fixed up your test asserts, found the issue was with self-referential but non-reflexive relationships (as you setup).  Added fix.

danielspaniel commented:
Your getting ever so very close to fixing all this @runspired .. you are still hanging in there which is pretty awesome. 

ok .. how's about this for some philosophical ramblings. 

 lets say we have polymorphic hats and the data is like this: 

```javascript
  const bigHatData = { data: { id: 1, type: "big-hat", attributes: { type: "big-hat" }, relationships: { "user": { "data": { "type": "user", "id": "1" } } } } };
  const smallHatData = { data: { id: 1, type: "small-hat", attributes: { type: "small-hat" }, relationships: { "user": { "data": { "type": "user", "id": "1" } } } } };
```
 
 upon close inspection one sees that the two models have the same id ( 1 ) but they are different sub-types. 

 If you push the user and then these two hats .. the test to get the user hats fails ( only finds 1 .. not 2 ) 

 yours/our tests are doing this 

```javascript 
  const bigHatData = { data: { id: 1, type: "big-hat", attributes: { type: "big-hat" }, relationships: { "user": { "data": { "type": "user", "id": "1" } } } } };
  const smallHatData = { data: { id: 2, type: "small-hat", attributes: { type: "small-hat" }, relationships: { "user": { "data": { "type": "user", "id": "1" } } } } };
```
  using a different id for every subtype   and therefore they pass .. but I think that the first way should pass as well .. because it used to in the past and I think polymorphic models can have same id for different type 

 you want me to PR this failing test ?  

 or you want to give up now and get a free carton of fig newtons and whatever is behind curtain number 2

runspired commented:
This is a hard one because this case has actually never worked properly. It seems to work on the surface but has long plagued us as a source of bug reports.

danielspaniel commented:
I hear you, but I not sure what to tell the databases that are storing out the data this way .. since they going to laugh at me if I ask for different id's tomorrow

runspired commented:
@danielspaniel I would push some tests for this case, I will see what I can do but can't promise much atm.

runspired commented:
@danielspaniel just a thought, but worst case it would be pretty easy for you to rectify the ID conflicts client-side by normalizing them to `${modelName}-${id}` on the way in and then stripping the modelName while serializing on the way out.

danielspaniel commented:
I guess I could, but so would everyone else .. and currently without the lazy relationship thing,  if you just create stuff like that , it works .. since create bypasses the loading lazy thing-a-majigger.

runspired commented:
I think @igorT @stefanpenner and @hjdivad need to weigh in on our intent to support conflicting IDs on polymorphic type.

danielspaniel commented:
ok .. fair enough ..

danielspaniel commented:
just remember as you discuss the fate of the polymorphic universe that polymorphic types with same ids like 

 'big-hat' id: 1
 'small-hat' id: 1 

 Are supported in Ember Data and work fine ( I have been using polymorphic models like this for many years in ember-data and never had a problem. )  
 They work right now in ED 2.14+  but ONLY if you don't use lazy relationship thing-a-bob. 
 If you create the models with createRecord it works ( in other words )

runspired commented:
@danielspaniel my point was not that they don't sometimes work today, just that they don't consistently work and have been a source of many bug reports.  A very simple error case is the following:

```js
/*
  if the request returns an entry of type `hat`,
  we end up with two identity map entries and the
  next request for findRecord or a call to `peekRecord`
  will return an empty model.
*/
store.findRecord('big-hat', '1')
```

Additional bugs include any relationship that has both a `big-hat` and `small-hat` with the same ID will via most entry-points end up with only one of the two entries.

danielspaniel commented:
That particular example is interesting @runspired . maybe that case where they seek to find a subtype should assert an error if it returns a different subtype ( one possibility ) 

I think it is just going to be hard to drop supporting same id's in subtypes because polymorphism is either single table or multi table and multi table will have same id's. How can you tell people to change their database  tables in order to use ember data? 

Seems like the best thing would be to enforce proper usage as I was suggesting above. Though I don't know all the cases that are causing the current meltdowns. 

Maybe ember data has to change the way it does things ( holds it's data ) in the long run to avoid these kinds of bugs, but the issue will always be there since people will still be using multi table polymorphic records. 

Just dropping the support in lazy relationships and keeping it elsewhere ( createRecord ) seems like not the greatest idea, but I know full well the frustration you are facing in trying to fix up something for a feature that might change in the future.

runspired commented:
@danielspaniel if we agree that for now you can make the IDs not conflict fairly easily on the client-side but also that we will continue to work on a solution to make multi-table work, would that be acceptable? In addition to a general desire to have a good solution, I likely have a project coming up soon that will require work on ember-data in this area.

danielspaniel commented:
I can't do the client side id munging you suggested for my application, so I would vote that down. But it's also others as well who have the same problem. You would force all ember-data users to do some hack. 

 All just to save the lazy relationship thing?  Why are you trying so hard to save that feature?  I think you're not a quitter and admire you for it.  But removing that broken feature solves the problem immediately.  Then we can work on ember data for the future to shore up the polymorphic issues ( which I am both down and up for )

The lazy relationships feature just adds another point of failure to ember-data and might be better off if we sent it to a remote island to live out it's days surfing and getting drunk.

danielspaniel commented:
But maybe there is a compromise?  
 
If the lazy relationships could somehow be made optional than that would work as an interim solution. 
If there could be environment setting or a way to set that in an initializer ( set to run before ember-data fires its engines ) to say =>  lazy-relationships: false     I think everyone would be happy.

danielspaniel commented:
I am not usually one to get excited about anything except grass growing in my lawn, but thanks for starting the  WIP to get to a conclusion on being able to support multi table inheritance.  Very very exciting.  👯

danielspaniel commented:
This branch now completes the fixes required for it to work with polymorphic and regular models. I see no other issues and the tests have solid coverage.

danielspaniel commented:
Can I help in any way to resolve this issue ( sometime soon .. by end of year? )

runspired commented:
@danielspaniel we discussed on the weekly ember-data call today ensuring that this is properly reviewed and landed before we release the next major, and that we will attempt to backport all the way to when lazy-relationships first landed.  Effectively this means that we expect to land and release this within the next week.

danielspaniel commented:
super .. thanks for update

igorT commented on addon/-private/system/relationships/relationship-payloads-manager.js:
```diff
@@ -172,7 +172,28 @@ export default class RelationshipPayloadsManager {
-    return this._cache[key];
+    let cache = this._cache[key];
+
+    if (cache === undefined) {
```
please add comments here as to what we are doing

igorT commented on addon/-private/system/relationships/relationship-payloads.js:
```diff
@@ -172,14 +235,17 @@ export default class RelationshipPayloads {
       let inverseIsMany;
+
+      const payloadId = this.payloadId(modelName, id);
+
```
Can you key on the type id pair instead of just the id please

igorT commented on addon/-private/system/relationships/relationship-payloads.js:
```diff
@@ -1,22 +1,56 @@
 /**
-  Manages the payloads for both sides of a single relationship, across all model
-  instances.
+ Manages the payloads for both sides of a single relationship, across all model
```
Are the whitespace changes on purpose?

BryanCrotaz commented:
Does this overlap with my PR #5268? It seems as though we're fixing the same problem in very different ways.

runspired commented:
@bryancrotaz 100% overlap

BryanCrotaz commented:
My solution seems simpler but then I think I must therefore be missing something...

runspired commented:
Check the various test cases added :P

runspired commented on addon/-private/system/relationships/relationship-payloads.js:
```diff
@@ -79,71 +91,91 @@ export default class RelationshipPayloads {
 
     if (this._isLHS(modelName, relationshipName)) {
-      delete this._lhsPayloads[id];
+      delete this.lhs_payloads.delete(modelName, id);
```
this is a mistake?

runspired replied:
seems sketch this passes tests

runspired commented on addon/-private/system/relationships/relationship-payloads.js:
```diff
@@ -79,71 +91,91 @@ export default class RelationshipPayloads {
-  */
+   @method
+   */
   _isLHS(modelName, relationshipName) {
```
Can you please rewrite/add to the comments here

runspired commented on addon/-private/system/relationships/relationship-payloads.js:
```diff
@@ -79,71 +91,91 @@ export default class RelationshipPayloads {
+    if (isRelationship === true) {
+      return isSelfReferential === true || // itself
+        modelName === relInfo.lhs_baseModelName || // base or non-polymorphic
+        relInfo.lhs_modelNames.indexOf(modelName) !== -1; // polymorphic
```
please add a comment here

danielspaniel commented:
I guess this never made it to 2.18 .. any idea when this might get merged ?

cibernox commented:
I believe I'm hitting this. Is there any more work to be done?

BryanCrotaz commented:
Has this stalled? What’s left to do? Can we all help?

bmac commented:
I've back-ported this fix to `2.14.11`, `2.15.4`, `2.16.4`, `2.17.1`, `2.18.2`, and `3.0.2`.

danielspaniel commented:
This was a good weekend. Thanks for the merge and backports.
